packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,555 @@
+    bool nameMatchesPackage(const std::vector<core::NameRef> &revNameParts) {
+        const auto &packageName = this->package.fullName();
+
+        if (revNameParts.size() < packageName.size()) {
+            return false;
+        }
+
+        return std::equal(packageName.begin(), packageName.end(), revNameParts.rbegin());
+    }
+
+public:
+    const core::packages::PackageInfo &package;
+    const bool insideTestFile;
+
+    VisibilityCheckerPass(core::Context ctx, const core::packages::PackageInfo &package)
+        : package{package}, insideTestFile{ctx.file.data(ctx).isPackagedTest()} {}
+
+    // `keep-def` will reference constants in a way that looks like a packaging violation, but is actually fine. This
+    // boolean allows for an early exit when we know we're in the context of processing one of these sends. Currently
+    // the only sends that we process this way will not have any nested method calls, but if that changes this will need
+    // to become a stack.
+    bool ignoreConstant = false;
+
+    ast::ExpressionPtr preTransformSend(core::Context ctx, ast::ExpressionPtr tree) {
+        auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
+        this->ignoreConstant = send.fun == core::Names::keepForIde();
+        return tree;
+    }
+
+    ast::ExpressionPtr postTransformSend(core::Context ctx, ast::ExpressionPtr tree) {
+        this->ignoreConstant = false;
+        return tree;
+    }
+
+    ast::ExpressionPtr postTransformConstantLit(core::Context ctx, ast::ExpressionPtr tree) {
+        if (this->ignoreConstant) {
+            return tree;
+        }
+
+        auto &lit = ast::cast_tree_nonnull<ast::ConstantLit>(tree);
+        if (!lit.symbol.isClassOrModule() && !lit.symbol.isFieldOrStaticField()) {
+            return tree;
+        }
+
+        auto loc = lit.symbol.loc(ctx);
+
+        auto otherFile = loc.file();
+        if (!otherFile.exists() || !otherFile.data(ctx).isPackaged()) {
+            return tree;
+        }
+
+        // If the imported symbol comes from the test namespace, we must also be in the test namespace.
+        if (otherFile.data(ctx).isPackagedTest() && !this->insideTestFile) {
+            if (auto e = ctx.beginError(lit.loc, core::errors::Packager::UsedTestOnlyName)) {
+                e.setHeader("Used test-only constant `{}` in non-test file", lit.symbol.show(ctx));
+            }
+        }
+
+        // no need to check visibility for these cases
+        auto otherPackage = ctx.state.packageDB().getPackageNameForFile(otherFile);
+        if (!otherPackage.exists() || this->package.mangledName() == otherPackage) {
+            return tree;
+        }
+
+        // Did we resolve a packaged symbol in unpackaged code?
+        if (!this->package.exists()) {
+            if (auto e = ctx.beginError(lit.loc, core::errors::Packager::PackagedSymbolInUnpackagedContext)) {
+                e.setHeader("Packaged constant `{}` used in an unpackaged context", lit.symbol.show(ctx));
```

Comment:
Do we have any tests that mention `# packaged: false` other than the one unit test in `core`?

Reply:
We don't, but I can add one to a packager test as it would be a nice way of also asserting the error about using packaged symbols from an unpackaged context.

Reply:
The behavior of the current packager is to allow references to packaged code from unpackaged contexts, as unpackaged contexts are defined with a cbase. I think that to stay compatible and make it possible to land this PR, we should remove this error for now, and add a TODO to come back to enforcing this.